### ISSUE TITLE
Query optimization in get_for_object_reference

### ISSUE THREAD

**Issue by lorinkoz** [etianen/django-reversion#430]
In `revisions.py`, `RevisionManager`, `get_for_object_reference`, line 517 you are applying a `select_related` to `revision`. Could it be possible to apply a `selected_related` also to `revision__user`, and perhaps a `prefetch_related` to the generic relation itself?
Since `get_unique_for_object` uses `get_for_object` (which in turn uses `get_for_object_reference`) but returns a list instead of a queryset (for obvious reasons), optimizing on the project's side is impossible, and it's very common to fall into a query+1 antipattern.
I am displaying in a view unique revisons with their related user, and just by adding the `select_related` to `revision__user` the number of queries were reduced by half.
I can submit a patch for it.

**Comment by etianen** [comment]
You can just add that to your own view logic. get_for_object_reference() just returns a queryset, so you can apply whatever optimizations you require to it.

**Comment by lorinkoz** [comment]
But I'm using `get_unique_for_object`, and I shouldn't re-implement that logic.

**Comment by etianen** [comment]
Ah, I see!

I think maybe the solution is to split out the de-duping logic into another function that takes a queryset, allowing it to be applied to any arbitrary queryset.

**Comment by lorinkoz** [comment]
Do you think it would be more convenient to create a custom queryset with a custom de-dup method that returns a list, and use that queryset instead?

**Comment by etianen** [comment]
That's a neat idea.

I think the de dupe code needs a revisit anyway. I suspect it can be done more efficiently as an iterator.

**Comment by lorinkoz** [comment]
Let me know if you're doing it yourself or if I should make a pull request.

**Comment by etianen** [comment]
I'll take a look at this on Tuesday, next week.

**Comment by lorinkoz** [comment]
Also, if you are willing to accept hitting the database twice, there is no need to return a list at all:
``` 
dup_list = []
last_version = None
for version in reversion.get_for_object(target_object) \
						.order_by('serialized_data', '-pk'):
	if last_version and last_version.serialized_data == version.serialized_data:
		dup_list.append(version.pk)
	last_version = version
versions = reversion.get_for_object(target_object) \
					.exclude(pk__in=dup_list)    
```
This same logic could be used for a management command in order to remove all duplicates in database. If you order by `serialized_data` you can even run the logic against all Version instances, no matter what object they are related to.

**Comment by etianen** [comment]
I was actually hoping to do this without loading all objects into memory by using queryset.iterator(), and returning a generator. There's no real advantage to returning a queryset if you can run the de-duping over any queryset, since you just perform any additional filtering first.

**Comment by lorinkoz** [comment]
In my opinion, it's always a good idea to return querysets, so that further filtering may be applied at the user discretion.

As for using `queryset.iterator`, it only makes sense for intermediate queries like the one collecting the duplicates. Django docs say that `iterator` won't work well with `prefetch_related` because they cancel each other (`iterator` prevents caching while `prefetch_related` cache results). Therefore, it makes no sense to prevent further caching (or deactivate previous anti-caching behaviour) on querysets being returned by your methods.

This is mainly motivated by an actual example in a project I'm working on: I changed the code for installed `django-reversion`. Am currently using a `select_related` on `revision` and `revision__user`, and a `prefetch_related` on `object`, it drastically reduces db overload when rendering a view with many (many!) versions that shows details about the user and the related object.

**Comment by etianen** [comment]
I see the rational. However, in the case of a db query where the number of objects in the queryset exceed 999, SQLite will break. Even on other db backends, huge IN clauses come with undesirable performance implications.

In addition, the memory usage will start to go high, so an iterator is generally what you want to do. Better to have something go slower, than to simply stop working above a certain size.

You can implement the two-query approach very easily on the result of the iterator, in any case, if it suits your purpose.

**Comment by etianen** [comment]
Fixed in #55e3709d730d06ae933ca557d0d7d704f2b05820
